lib/std/Thread/Futex.zig

```diff
@@ -21,160 +19,149 @@ const spinLoopHint = std.atomic.spinLoopHint;
+pub fn timedWait(ptr: *const Atomic(u32), expect: u32, timeout_ns: u64) error{Timeout}!void {
+    @setCold(true);
 
-    // Avoid calling into the OS for no-op waits()
-    if (timeout) |timeout_ns| {
-        if (timeout_ns == 0) {
-            if (ptr.load(.SeqCst) != expect) return;
-            return error.TimedOut;
+    // Avoid calling into the OS for no-op timeouts.
+    if (timeout_ns == 0) {
+        if (ptr.load(.SeqCst) != expect) {
+            return error.Timeout;
         }
     }
 
-    return OsFutex.wait(ptr, expect, timeout);
+    return Impl.wait(ptr, expect, timeout_ns);
 }
 
-/// Unblocks at most `num_waiters` callers blocked in a `wait()` call on `ptr`.
-/// `num_waiters` of 1 unblocks at most one `wait(ptr, ...)` and `maxInt(u32)` unblocks effectively all `wait(ptr, ...)`.
-pub fn wake(ptr: *const Atomic(u32), num_waiters: u32) void {
-    if (single_threaded) return;
-    if (num_waiters == 0) return;
+/// Unblocks at most `max_waiters` callers blocked in a `wait()` call on `ptr`.
+pub fn wake(ptr: *const Atomic(u32), max_waiters: u32) void {
+    @setCold(true);
 
-    return OsFutex.wake(ptr, num_waiters);
+    // Avoid calling into the OS if there's nothing to wake up.
+    if (max_waiters == 0) {
+        return;
+    }
+
+    Impl.wake(ptr, max_waiters);
 }
 
-const OsFutex = if (target.os.tag == .windows)
-    WindowsFutex
-else if (target.os.tag == .linux)
-    LinuxFutex
-else if (target.isDarwin())
-    DarwinFutex
-else if (builtin.link_libc)
-    PosixFutex
+const Impl = if (builtin.single_threaded)
+    SerialImpl
+else if (builtin.os.tag == .windows)
+    WindowsImpl
+else if (builtin.os.tag.isDarwin())
+    DarwinImpl
+else if (builtin.os.tag == .linux)
+    LinuxImpl
+else if (builtin.os.tag == .freebsd)
+    FreebsdImpl
+else if (builtin.os.tag == .openbsd)
+    OpenbsdImpl
+else if (builtin.os.tag == .dragonfly)
+    DragonflyImpl
+else if (std.Thread.use_pthreads)
+    PosixImpl
 else
-    UnsupportedFutex;
+    UnsupportedImpl;
 
-const UnsupportedFutex = struct {
-    fn wait(ptr: *const Atomic(u32), expect: u32, timeout: ?u64) error{TimedOut}!void {
+/// We can't do @compileError() in the `Impl` switch statement above as its eagerly evaluated.
+/// So instead, we @compileError() on the methods themselves for platforms which don't support futex.
+const UnsupportedImpl = struct {
+    fn wait(ptr: *const Atomic(u32), expect: u32, timeout: ?u64) error{Timeout}!void {
         return unsupported(.{ ptr, expect, timeout });
     }
 
-    fn wake(ptr: *const Atomic(u32), num_waiters: u32) void {
-        return unsupported(.{ ptr, num_waiters });
+    fn wake(ptr: *const Atomic(u32), max_waiters: u32) void {
+        return unsupported(.{ ptr, max_waiters });
     }
 
     fn unsupported(unused: anytype) noreturn {
-        @compileLog("Unsupported operating system", target.os.tag);
+        @compileLog("Unsupported operating system", builtin.target.os.tag);
```

Comment:
Let's use the language features as they were intended to be used:

```suggestion
        @compileError("Unable to provide std.Thread.Futex implementation for target OS " ++ @tagName(builtin.target.os.tag));
```

And then on the side, let's discuss why `@compileLog` is more attractive for this use case than `@compileError` because the very fact that you are tempted to use the unintended builtin here means the language has a design flaw.

Reply:
`@compileLog` is only more attractive due to having built-in varargs and serialization. Not sure if this could (or even should) be added to `@compileError`. Either way, it's a quick fix.